### PR TITLE
Add billing limiter

### PR DESCRIPTION
**Pull request by ellie**
I decided against the background worker for reading from Redis. We can always implement this in the future, but it introduced less predictable locking behaviour that I'd rather avoid unless necessary. 

This implementation will, at worst, slow a request down by 10 ms. Any more than 10ms to read from redis will timeout, and fail open. Checking our telemetry shows that almost all of our redis requests complete in much less than 10ms, and zrangebyscore in <=1ms

The limiter will compare the token to a cached set of limited tokens. Every `interval`, we fetch the latest set from redis and store it in a cache.

Each request is compared against the cache - if the token matches, we return 200 without any further processing. This is unfortunate, but matches the current behaviour.

Our v1 endpoint (not in this PR), will return 429 with some context.

### CODE REVIEW

File: capture/src/billing_limits.rs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        conn.zrangebyscore(format!("{QUOTA_LIMITER_CACHE_KEY}events"), now, "+Inf").await
+
+    }
+
+    /// Spawn the tokio updater task
+    ///
+    /// Note: If no nodes are set, the billing limiter falls back to limiting a fixed set. No
+    /// updates.
+    ///
+    /// This will return an error if called more than once, globally. The background task will run
```
Comment (ellie) [inline capture/src/billing_limits.rs:77]: Not presently true

File: capture/src/billing_limits.rs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        limiter
+            .start_updater()
+            .await
+            .expect("Failed to start updater");
+
+        assert_eq!(limiter.is_limited("idk it doesn't matter").await, false);
+        assert!(limiter.is_limited("some_org_hit_limits").await);
+    }
+
+    // TODO: test the redis stuff somehow
```
Comment (ellie) [inline capture/src/billing_limits.rs:180]: honestly I'm probably just going to wrap the client in my own trait, so I can inject failures and make sure we handle them right

Comment (ellie) [reply capture/src/billing_limits.rs:180]: I really wanted to use `ConnectionLike` but it got kinda weird

File: capture/src/billing_limits.rs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                        l.clear();
+                    }
+                }
+            }
+        });
+
+        Ok(())
+    }
+
+    pub async fn is_limited(&self, key: &str) -> bool {
```
Comment (xvello) [inline capture/src/billing_limits.rs:130]: Although the ideal topology will be to have a dedicated capture pool per data type, we won't have that luxury on local devenv & hobby. Plus we might decide to colocate several "beta" data types in a common pool, to reduce the cost of an MVP.

For that, we'll want a `data_type` enum, that is a parameter of that function. I think it's fine to sync all of `QUOTA_LIMITER_CACHE_KEY` for now for simplicity

File: capture/src/billing_limits.rs [hunk 4]
```diff
@@ -0,0 +1,181 @@
+use std::{collections::HashSet, sync::Arc};
+
+use async_trait::async_trait;
+/// Limit accounts by team ID if they hit a billing limit
+///
+/// We have an async celery worker that regularly checks on accounts + assesses if they are beyond
+/// a billing limit. If this is the case, a key is set in redis.
+///
+/// On the capture side, we should check if a request should be limited. If so, we can respond with
```
[the remark this answers is not in the document]
Comment (ellie) [reply capture/src/billing_limits.rs:10]: Totally makes sense - I've left it as `v0` behaviour for now, much as it makes me sad.

File: capture/src/billing_limits.rs (hunk 3 above)
Comment (ellie) [reply capture/src/billing_limits.rs]: done 👍  just copied it pretty much as is from the python

**Comment by ellie** [comment]
I've reverted from my initial plan of the background task. I think I was overcomplicating things - the implementation here should work well enough and we can always revisit if that is not the case.

File: capture/tests/billing_limiter.rs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    async fn send_batch(&self, events: Vec<ProcessedEvent>) -> Result<(), CaptureError> {
+        self.events.lock().unwrap().extend_from_slice(&events);
+        Ok(())
+    }
+}
+
+#[tokio::test]
+async fn it_matches_django_capture_behaviour() -> anyhow::Result<()> {
```
[the remark this answers is not in the document]
Comment (ellie) [reply capture/tests/billing_limiter.rs:76]: I did indeed - was going to test it end to end, but realised there's no external way of knowing if we're limited if we're just returning OK

**Review by xvello — APPROVED** [review]

**Pull request merged**